### PR TITLE
Sy 1170 remains true for block in control sequences

### PR DESCRIPTION
**Pull request by emilbon99**
# Feature Pull Request Template

## Key Information

- **Linear Issue**: [SY-1170](https://linear.app/synnax/issue/SY-1170/remains-true-for-block-in-control-sequences)

## Description

Implements [SY-1170](sy-1170-remains_true_for-block-in-control-sequences), [SY-1173](https://linear.app/synnax/issue/SY-1173/add-a-while-block-to-the-controller), and [SY-1164](https://linear.app/synnax/issue/SY-1164/supplement-controller-wait-conditions)

## Basic Readiness Checklist

- [x] I have performed a self-review of my code.
- [x] I have added relevant tests to cover the changes to CI.
- [x] I have updated user facing documentation accordingly.
- [x] I have verified code coverage targets are met.

## Migrations

- [ ] Console - I have ensured that previous versions of stored data structures are 
properly migrated to new formats.
- [ ] Server - I have ensured that previous versions of stored data structures are 
properly migrated to new formats.

##  Additional Notes
- [ ] These changes deal with concurrency.
- [ ] These changes affect UI.

## Manual QA Additions

- [ ] I have updated the [Release Candidate](/.github/PULL_REQUEST_TEMPLATE/rc.md) template
with necessary manual QA steps to test my change.

## Breaking Changes

Please list any breaking changes to public or internal packages.

## Reviwer Checklist
- [ ] Sufficient test coverage of new additions.
- [ ] Verified all steps in the Readiness checklists.
- [ ] UI changes have been tested.
- [ ] Style and formatting is consistent.
- [ ] Reviewed any relevant changes to concurrent code for safety. 
- [ ] Sufficient comments and clarity of code.
- [ ] Any additional documentation necessary for feature is provided and has been reviewed for clarity.

### CODE REVIEW

File: client/py/synnax/timing/timing.py [hunk 1]
```diff
@@ -45,7 +42,116 @@ def _precise_sleep(dur: TimeSpan | float | int):
...
 
 
-def sleep(dur: TimeSpan | float | int, precise: bool = False):
+def sleep(dur: Rate | TimeSpan | float | int, precise: bool = False):
+    """Sleep is a function that sleeps for the given duration. The duration can be
+    specified as a Rate, TimeSpan, float, or int. If the precise flag is set to True,
+    the function will use a more precise sleep implementation. It uses welford's
+    algorithm to estimate the ideal time to sleep for the given duration. This function
+    uses considerably more CPU than time.sleep, so it should only be used when high
+    precision is required.
```
Comment (pjdotson) [inline client/py/synnax/timing/timing.py:51]: This docstring is somewhat unclear - when `precise` is `False`, it uses the same CPU time as time.sleep. I would do two things:
1. Make clear that when `dur` is a `float` or `int`, the time is in seconds.
2. Move the commentary on the precise sleep to a second paragraph to be clear that the CPU time and Welford's algorithm comment are just for precise sleep
3. Capitalize Welford.

File: client/py/synnax/timing/timing.py [hunk 2]
```diff
@@ -10,19 +10,16 @@
...
+RESOLUTION = (100 * TimeSpan.MICROSECOND).seconds
 
-
-def _precise_sleep(dur: TimeSpan | float | int):
+def _precise_sleep(dur: float | int):
     """Sleep implements a higher precision alternative to time.sleep. It uses welford's
     algorithm to estimate the ideal time to sleep for the given duration. This function
     uses considerably more CPU than time.sleep, so it should only be used when high
     precision is required.
     """
```
Comment (pjdotson) [inline client/py/synnax/timing/timing.py:22]: You can delete this docstring.

File: client/py/synnax/timing/timing.py [hunk 3]
```diff
@@ -45,7 +42,116 @@ def _precise_sleep(dur: TimeSpan | float | int):
...
+    counter to measure time. The timer can be started, reset, and the elapsed time can
+    be queried.
+    """
+    _start: TimeSpan
+
+    def __init__(self):
+        self.reset()
+
+    def elapsed(self) -> TimeSpan:
+        """Elapsed returns the time elapsed since the timer was started.
```
Comment (pjdotson) [inline client/py/synnax/timing/timing.py:77]: lowercase `elapsed`.

File: client/py/synnax/timing/timing.py [hunk 4]
```diff
@@ -45,7 +42,116 @@ def _precise_sleep(dur: TimeSpan | float | int):
...
+        """
+        return TimeSpan(time.perf_counter_ns() - self._start)
+
+    def start(self):
+        """Start starts the timer.
```
Comment (pjdotson) [inline client/py/synnax/timing/timing.py:82]: Lowercase `start`.

File: client/py/synnax/timing/timing.py [hunk 5]
```diff
@@ -45,7 +42,116 @@ def _precise_sleep(dur: TimeSpan | float | int):
...
+        """
+        self.reset()
+
+    def reset(self):
+        """Reset resets the timer to zero.
```
Comment (pjdotson) [inline client/py/synnax/timing/timing.py:87]: Lowercase `reset`.

File: client/py/synnax/timing/timing.py [hunk 6]
```diff
@@ -45,7 +42,116 @@ def _precise_sleep(dur: TimeSpan | float | int):
...
+    return time.sleep(dur)
+
+
+class Timer:
+    """Timer is a class that implements a simple timer. It is used to measure the time
+    elapsed since the timer was started. The timer uses the high-resolution performance
+    counter to measure time. The timer can be started, reset, and the elapsed time can
+    be queried.
+    """
+    _start: TimeSpan
```
Comment (pjdotson) [inline client/py/synnax/timing/timing.py:71]: Shouldn't `_start` be a `TimeStamp` and not a `TimeSpan`?

File: client/py/synnax/timing/timing.py [hunk 7]
```diff
@@ -45,7 +42,116 @@ def _precise_sleep(dur: TimeSpan | float | int):
...
+        self._start = TimeSpan(time.perf_counter_ns())
+
+
+class Loop:
+    """Loop is a class that implements a rate-limited loop. It is used to ensure that a
+    block of code is executed at a fixed rate. The loop will sleep for the remainder of
+    the interval if the block of code executes faster than the interval. If the block of
+    code takes longer than the interval, the loop will skip sleeping to ensure that the
+    block of code is executed as soon as possible. The loop keeps track of the average
+    execution time and uses this to correct for any drift in the interval.
```
Comment (pjdotson) [inline client/py/synnax/timing/timing.py:98]: Should also document the public class properties.

File: client/py/synnax/framer/adapter.py [hunk 8]
```diff
@@ -103,7 +104,7 @@ def __adapt_ch(
     ) -> ChannelPayload:
         if not isinstance(ch, (ChannelKey, ChannelName)):
             return ch
-        return retrieve_required(self.retriever, ch)[0]
+        return retrieve_one_required(self.retriever, ch)
```
Comment (pjdotson) [inline client/py/synnax/framer/adapter.py:107]: Personally I think that the logic flows clearer like this:

```python
if isinstance(ch, (ChannelKey, ChannelName)):
    return retrieve_one_required(self.retriever, ch)
return ch
```

File: client/py/examples/control/virtual.py
Comment (pjdotson) [inline client/py/examples/control/virtual.py:1]: Should this really go in the examples folder? It's not super clear what its doing, if you want to keep the code you should put it in the `dev` folder.

File: client/py/synnax/channel/retrieve.py [hunk 9]
```diff
@@ -70,6 +73,19 @@ def retrieve(self, params: ChannelParams) -> list[ChannelPayload]:
         req = _Request(**{normal.variant: normal.params})
         return send_required(self._client, _ENDPOINT, req, _Response).channels
 
+    @trace("debug")
+    def retrieve_one(self, param: ChannelKey | ChannelName) -> ChannelPayload | None:
```
Comment (pjdotson) [inline client/py/synnax/channel/retrieve.py:77]: Add test

File: client/py/synnax/channel/retrieve.py [hunk 10]
```diff
@@ -133,14 +149,25 @@ def _get(self, param: ChannelKey | ChannelName) -> list[ChannelPayload] | None:
                 channels.append(ch)
         return None if len(channels) == 0 else channels
 
+    def get_one(self, param: ChannelKey | ChannelName) -> ChannelPayload | None:
```
Comment (pjdotson) [inline client/py/synnax/channel/retrieve.py:152]: Add test

File: client/py/synnax/channel/retrieve.py [hunk 11]
```diff
@@ -133,14 +149,25 @@ def _get(self, param: ChannelKey | ChannelName) -> list[ChannelPayload] | None:
...
         for channel in channels:
-            self._channels[channel.key] = channel
-            keys = self._names_to_keys.get(channel.name)
-            if keys is None:
-                self._names_to_keys[channel.name] = {channel.key}
-            else:
-                keys.add(channel.key)
+            self.set_one(channel)
+
+    def set_one(self, channel: ChannelPayload) -> None:
```
Comment (pjdotson) [inline client/py/synnax/channel/retrieve.py:164]: Add test

File: client/py/synnax/channel/retrieve.py [hunk 12]
```diff
@@ -162,6 +189,14 @@ def retrieve(self, params: ChannelParams) -> list[ChannelPayload]:
         results.extend(retrieved)
         return results
 
+    def retrieve_one(self, param: ChannelKey | ChannelName) -> ChannelPayload | None:
```
Comment (pjdotson) [inline client/py/synnax/channel/retrieve.py:192]: Add test

File: client/py/synnax/channel/retrieve.py [hunk 13]
```diff
@@ -176,3 +211,11 @@ def retrieve_required(
     if len(not_found) > 0:
         raise NotFoundError(f"Could not find channels: {not_found}")
     return results
+
+def retrieve_one_required(
```
Comment (pjdotson) [inline client/py/synnax/channel/retrieve.py:215]: Add test

**Review by pjdotson — COMMENTED** [review]
Approve after changes are made.

File: client/py/synnax/channel/retrieve.py (hunk 11 above)
Comment (emilbon99) [reply client/py/synnax/channel/retrieve.py:164]: Made this internal

File: client/py/synnax/channel/retrieve.py (hunk 12 above)
Comment (emilbon99) [reply client/py/synnax/channel/retrieve.py:192]: This is a function that is only used internally, so it's not publicly exposed for testing. It gets hammered pretty hard by publicly available test cases.

Comment (emilbon99) [reply client/py/synnax/channel/retrieve.py:192]: Actually went in and added cases

**Pull request merged**